avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java

```diff
@@ -815,145 +811,151 @@ private Object toConnectData(Schema schema, Object value) {
-            Map<CharSequence, Object> result = new HashMap<>(original.size());
-            for (Map.Entry<CharSequence, Object> entry : original.entrySet()) {
-              result.put(entry.getKey().toString(),
-                         toConnectData(valueSchema, entry.getValue()));
-            }
-            converted = result;
-          } else {
-            // Arbitrary keys
-            List<IndexedRecord> original = (List<IndexedRecord>) value;
-            Map<Object, Object> result = new HashMap<>(original.size());
-            for (IndexedRecord entry : original) {
-              int avroKeyFieldIndex = entry.getSchema().getField(KEY_FIELD).pos();
-              int avroValueFieldIndex = entry.getSchema().getField(VALUE_FIELD).pos();
-              Object convertedKey = toConnectData(keySchema, entry.get(avroKeyFieldIndex));
-              Object convertedValue = toConnectData(valueSchema, entry.get(avroValueFieldIndex));
-              result.put(convertedKey, convertedValue);
-            }
-            converted = result;
-          }
-          break;
-        }
-
-        case STRUCT: {
-          // Special case support for union types
-          if (schema.name() != null && schema.name().equals(AVRO_TYPE_UNION)) {
-            Schema valueRecordSchema = null;
-            if (value instanceof IndexedRecord) {
-              IndexedRecord valueRecord = ((IndexedRecord) value);
-              valueRecordSchema = toConnectSchema(valueRecord.getSchema(), true, null, null);
-            }
-            for (Field field : schema.fields()) {
-              Schema fieldSchema = field.schema();
-
-              if (isInstanceOfAvroSchemaTypeForSimpleSchema(fieldSchema, value) ||
-                  (valueRecordSchema != null && valueRecordSchema.equals(fieldSchema))) {
-                converted = new Struct(schema).put(unionMemberFieldName(fieldSchema),
-                                                   toConnectData(fieldSchema, value));
-                break;
-              }
-            }
-            if (converted == null) {
-              throw new DataException(
-                  "Did not find matching union field for data: " + value.toString());
-            }
-          } else {
-            IndexedRecord original = (IndexedRecord) value;
-            Struct result = new Struct(schema);
-            for (Field field : schema.fields()) {
-              int avroFieldIndex = original.getSchema().getField(field.name()).pos();
-              Object convertedFieldValue
-                  = toConnectData(field.schema(), original.get(avroFieldIndex));
-              result.put(field, convertedFieldValue);
-            }
-            converted = result;
-          }
-          break;
-        }
-
-        default:
-          throw new DataException("Unknown Connect schema type: " + schema.type());
+      // @TODO: should this be handling when the schema isn't optional but value is null?
```

Comment:
Agreed that it would be nice to have a verbose error thrown here for that case rather than falling through to one of the cases below that'll result in an NPE.